### PR TITLE
PHOENIX-7166 : Enable feature flags introduced by metadata caching redesign

### PR DESCRIPTION
**Pull request by palashc**

### CODE REVIEW

**Comment by palashc** [comment]
@shahrs87 Let me know if we can get this one in now?

File: phoenix-core-client/src/main/java/org/apache/phoenix/execute/MutationState.java [hunk 1]
```diff
@@ -1240,6 +1241,15 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
                     String tableName = tableRef.getTable().getTableName().toString();
                     mc.updateCache(tenantId, schemaName, tableName, true);
                 }
+            } catch (MetaDataEntityNotFoundException e) {
```
Comment (shahrs87) [inline phoenix-core-client/src/main/java/org/apache/phoenix/execute/MutationState.java:1244]: Why would `ValidateLastDDLTimestampUtil#validateLastDDLTimestamp` throw `MetaDataEntityNotFoundException` ?

Comment (palashc) [reply phoenix-core-client/src/main/java/org/apache/phoenix/execute/MutationState.java]: Removed that change.

**Review by shahrs87 — CHANGES_REQUESTED** [review]
some changes.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpgradeIT.java [hunk 2]
```diff
@@ -731,13 +730,14 @@ public void testLastDDLTimestampBootstrap() throws Exception {
...
     private void nullDDLTimestamps(Connection conn) throws SQLException {
+        //ignore system tables since that can interfere with other tests.
         String pkCols = TENANT_ID + ", " + TABLE_SCHEM +
             ", " + TABLE_NAME + ", " + COLUMN_NAME + ", " + COLUMN_FAMILY;
         String upsertSql =
             "UPSERT INTO " + SYSTEM_CATALOG_NAME + " (" + pkCols + ", " +
                 LAST_DDL_TIMESTAMP + ")" + " " +
                 "SELECT " + pkCols + ", NULL FROM " + SYSTEM_CATALOG_NAME + " " +
-                "WHERE " + TABLE_TYPE + " IS NOT NULL";
+                "WHERE " + TABLE_TYPE + " " + " != '" + PTableType.SYSTEM.getSerializedValue() + "'";
```
[the remark this answers is not in the document]
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/UpgradeIT.java:740]: Tests in this class passed when run individually but fail when the whole test class is run. One test `testLastDDLTimestampBootstrap()` in the class was interfering with the other tests. That test nulls out all LAST_DDL_TSs for test purpose and did not ignore system tables. 

The change is in the helper method `nullDDLTimestamps` which nulls out all tables to ignore SYSTEM tables.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewIT.java [hunk 3]
```diff
@@ -660,6 +661,14 @@ public void testCreateChildViewWithBaseTableIndex(boolean localIndex)
                 fail();
             } catch (Exception ignore) {
             }
+
+            PTable view = PhoenixRuntime.getTable(conn, fullViewName);
```
[the remark this answers is not in the document]
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewIT.java:668]: I can not seem to remember why I added these lines, those PTable variables are not even accessed later. I will remove this change.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpdateCacheAcrossDifferentClientsIT.java [hunk 4]
```diff
@@ -104,17 +105,19 @@ public void testUpdateCacheFrequencyWithAddAndDropTable() throws Exception {
             } catch (TableNotFoundException e) {
                 //Expected
             }
-            rs = conn2.createStatement().executeQuery("select * from "+tableName);
             try {
+                rs = conn2.createStatement().executeQuery("select * from "+tableName);
                 rs.next();
                 fail("Should throw org.apache.hadoop.hbase.TableNotFoundException since the latest metadata " +
                         "wasn't fetched");
-            } catch (PhoenixIOException ex) {
```
[the remark this answers is not in the document]
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/UpdateCacheAcrossDifferentClientsIT.java:112]: client-1 dropped the table, client-2 will get TableNotFoundException i.e. SQLException when validating last ddl timestamp.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java [hunk 5]
```diff
@@ -1149,10 +1152,8 @@ public void testDroppingIndexedColDropsViewIndex() throws Exception {
             byte[] viewIndexPhysicalTable = viewIndex.getPhysicalName().getBytes();
             assertNotNull("Can't find view index", viewIndex);
             assertEquals("Unexpected number of indexes ", 2, view.getIndexes().size());
-            assertEquals("Unexpected index ",  fullNameViewIndex1 , view.getIndexes().get(0).getName()
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: Why is this change needed?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java [hunk 6]
```diff
@@ -1174,21 +1175,16 @@ public void testDroppingIndexedColDropsViewIndex() throws Exception {
             }
             
             // verify index metadata was dropped
-            try {
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1177]: why is this change required?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java [hunk 7]
```diff
@@ -1174,21 +1175,16 @@ public void testDroppingIndexedColDropsViewIndex() throws Exception {
...
-            
             pconn = viewConn.unwrap(PhoenixConnection.class);
-            view = pconn.getTable(new PTableKey(tenantId,  viewOfTable ));
+            view = pconn.getTableNoCache(viewOfTable);
+            assertEquals("Unexpected number of indexes ", 1, view.getIndexes().size());
+            assertEquals("Unexpected index ",  fullNameViewIndex2 , view.getIndexes().get(0).getName().getString());
+            assertNotEquals("Dropped index should not be in view metadata ",  fullNameViewIndex1 , view.getIndexes().get(0).getName().getString());
             try {
-                viewIndex = pconn.getTable(new PTableKey(tenantId,  fullNameViewIndex1 ));
+                viewIndex = pconn.getTableNoCache(fullNameViewIndex1);
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1184]: why is this change required?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameBaseIT.java [hunk 8]
```diff
@@ -530,13 +531,16 @@ protected void validateIndex(Connection connection, String tableName, boolean is
     }
 
     public static void renameAndDropPhysicalTable(Connection conn, String tenantId, String schema, String tableName, String physicalName, boolean isNamespaceEnabled) throws Exception {
+        // if client is validating last_ddl_timestamp, this change in physical table name should be visible to the client
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameBaseIT.java:534]: Do we need this change given that default UCF is set to ALWAYS?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameExtendedIT.java [hunk 9]
```diff
@@ -209,21 +209,26 @@ public void testHint() throws Exception {
             createIndexOnTable(conn, tableName, indexName2);
             populateTable(conn, tableName, 1, 2);
 
-            // Test hint
             String tableSelect = "SELECT V1,V2,V3 FROM " + tableName;
             ResultSet rs = conn.createStatement().executeQuery("EXPLAIN " + tableSelect);
-            assertEquals(true, QueryUtil.getExplainPlan(rs).contains(indexName));
+            String plan = QueryUtil.getExplainPlan(rs);
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameExtendedIT.java:214]: why is this change required?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java [hunk 10]
```diff
@@ -1173,10 +1174,10 @@ private void helpTestQueryForViewOnTableThatHasIndex(Statement s1,
         s1.execute("create index " + indexName + " ON " + tableName
                 + " (col2)");
 
-        try (ResultSet rs = s2.executeQuery("explain select /*+ INDEX("
+        try (ResultSet rs = s2.executeQuery("select /*+ INDEX("
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:1177]: why this change?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexAsyncThresholdIT.java [hunk 11]
```diff
@@ -162,9 +162,7 @@ public void testAsyncIndexCreation() throws Exception {
                 exception = (SQLException) e;
             }
             connection.commit();
-            PTableKey key = new PTableKey(null, this.tableName);
-            PMetaData metaCache = connection.unwrap(PhoenixConnection.class).getMetaDataCache();
-            List<PTable> indexes = metaCache.getTableRef(key).getTable().getIndexes();
+            List<PTable> indexes = connection.unwrap(PhoenixConnection.class).getTable(this.tableName).getIndexes();
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexAsyncThresholdIT.java:165]: why this change?

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java [hunk 12]
```diff
@@ -1276,14 +1281,24 @@ public void testMetricsWithIndexUsage() throws Exception {
                 assertTrue(metricExists);
                 metricExists = false;
                 //assert BaseTable is not being queried
-                for (PhoenixTableMetric metric : getPhoenixTableClientMetrics().get(dataTable)) {
-                    if (metric.getMetricType().equals(SELECT_SQL_COUNTER)) {
-                        metricExists = true;
-                        assertMetricValue(metric, SELECT_SQL_COUNTER, 0, CompareOp.EQ);
-                        break;
+                //if client is validating last_ddl_timestamps with ucf=never,
+                //there will be no metrics for base table (like getTable RPC times/counts).
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java:1285]: Can you explain more on this?

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java [hunk 13]
```diff
@@ -142,18 +146,25 @@ public void testSystemTablesAreInCache() throws Exception {
      */
     @Test
     public void testGlobalClientCacheMetrics() throws Exception {
-        int numThreads = 5;
+        int numThreads = 1;
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:149]: why this change?

File: phoenix-core/src/test/java/org/apache/phoenix/jdbc/ParallelPhoenixConnectionFailureTest.java [hunk 14]
```diff
@@ -51,9 +52,12 @@ public class ParallelPhoenixConnectionFailureTest extends BaseTest {
     @Test
     public void testExecuteQueryChainFailure() throws SQLException {
         HBaseTestingUtility hbaseTestingUtility = new HBaseTestingUtility();
-
-        PhoenixConnection conn1 = (PhoenixConnection) DriverManager.getConnection(url);
-        PhoenixConnection conn2 = (PhoenixConnection) DriverManager.getConnection(url);
+        Properties props = new Properties();
+        // disable last_ddl_timestamp validation for tests which use ConnectionLessQueryService
+        // since there is no hbase connection to get list of live region servers
+        props.put(QueryServices.LAST_DDL_TIMESTAMP_VALIDATION_ENABLED, Boolean.toString(false));
```
Comment (shahrs87) [inline phoenix-core/src/test/java/org/apache/phoenix/jdbc/ParallelPhoenixConnectionFailureTest.java:58]: The default is false. Do we need to set it to false again?

File: phoenix-core/src/test/java/org/apache/phoenix/query/BaseConnectionlessQueryTest.java [hunk 15]
```diff
@@ -86,7 +88,11 @@ private static void startServer(String url) throws Exception {
         // only load the test driver if we are testing locally - for integration tests, we want to
         // test on a wider scale
         if (PhoenixEmbeddedDriver.isTestUrl(url)) {
-            driver = initDriver(ReadOnlyProps.EMPTY_PROPS);
+            Map<String, String> props = Maps.newHashMapWithExpectedSize(1);
+            // disable last_ddl_timestamp validation for tests which use ConnectionLessQueryService
+            // since there is no hbase connection to get list of live region servers
+            props.put(QueryServices.LAST_DDL_TIMESTAMP_VALIDATION_ENABLED, Boolean.toString(false));
```
Comment (shahrs87) [inline phoenix-core/src/test/java/org/apache/phoenix/query/BaseConnectionlessQueryTest.java:94]: The default is false. Do we need to set it to false again?

**Review by shahrs87 — CHANGES_REQUESTED** [review]
Some comments.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java (hunk 5 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: The asserts here assume a certain ordering of indexes in the view's PTable but that ordering does not hold if the view PTable was refreshed in the client's cache. So I changed the asserts to look for the index name in a list of expected index names.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java (hunk 7 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1184]: This is to make sure that the view index is fetched from the server and not from the client's cache. The test is verifying whether the view index was dropped on the server.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameBaseIT.java (hunk 8 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameBaseIT.java:534]: We will have to make the change in our light fork then. Keeping this change does not affect the test even with UCF=always.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameExtendedIT.java (hunk 9 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/LogicalTableNameExtendedIT.java:214]: There is assumption on the ordering of index PTables inside the parent table PTable. Optimizer can pick any of them but the test assumes only the first one in the list will get picked.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexAsyncThresholdIT.java (hunk 11 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/index/IndexAsyncThresholdIT.java:165]: This is to make sure we fetch the latest ptable. I think such changes are okay to keep even if UCF=always.

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java (hunk 12 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java:1285]: Since the query plan picks the index, client never interacts with the base table directly if UCF=never and hence metrics related objects are not initialized.

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java (hunk 13 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:149]: not sure I remember this one, I will revert this.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/LoadSystemTableSnapshotIT.java [hunk 16]
```diff
@@ -167,6 +167,8 @@ public void testPhoenixUpgrade() throws Exception {
         serverProps.put(QueryServices.IS_NAMESPACE_MAPPING_ENABLED, "true");
         Map<String, String> clientProps = Maps.newHashMapWithExpectedSize(2);
         clientProps.put(QueryServices.IS_NAMESPACE_MAPPING_ENABLED, "true");
+        clientProps.put(QueryServices.DEFAULT_UPDATE_CACHE_FREQUENCY_ATRRIB, Long.toString(0L));
```
Comment (palashc) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/LoadSystemTableSnapshotIT.java:170]: Remove this

File: phoenix-core/src/it/java/org/apache/phoenix/jdbc/HighAvailabilityTestingUtility.java [hunk 17]
```diff
@@ -536,6 +539,12 @@ private static void setUpDefaultHBaseConfig(Configuration conf) {
 
             // Hadoop cluster settings to avoid failing tests
             conf.setInt(DFS_REPLICATION_KEY, 1); // we only need one replica for testing
+
+            // Phoenix Region Server Endpoint needed for metadata caching
+            if (QueryServicesOptions.DEFAULT_LAST_DDL_TIMESTAMP_VALIDATION_ENABLED) {
```
Comment (palashc) [inline phoenix-core/src/it/java/org/apache/phoenix/jdbc/HighAvailabilityTestingUtility.java:544]: remove the if check

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java (hunk 5 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: Need to look into this test

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java (hunk 10 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:1177]: will look into this

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpdateCacheAcrossDifferentClientsIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/UpdateCacheAcrossDifferentClientsIT.java:112]: @shahrs87 The difference comes from where the exception is thrown. 

- The test has given UCF=1000000000 when creating the table. So when client-2 from old design tried to query the dropped table, it does not refresh its cache. `executeQuery` does not error out. `rs.next()` throws `PhoenixIOException`
```
org.apache.phoenix.exception.PhoenixIOException: org.apache.phoenix.exception.PhoenixIOException: N000001
	at org.apache.phoenix.util.ClientUtil.parseServerException(ClientUtil.java:70)
	at org.apache.phoenix.iterate.BaseResultIterators.getIterators(BaseResultIterators.java:1464)
	at org.apache.phoenix.iterate.BaseResultIterators.getIterators(BaseResultIterators.java:1398)
	at org.apache.phoenix.iterate.RoundRobinResultIterator.getIterators(RoundRobinResultIterator.java:188)
	at org.apache.phoenix.iterate.RoundRobinResultIterator.next(RoundRobinResultIterator.java:94)
	at org.apache.phoenix.jdbc.PhoenixResultSet.next(PhoenixResultSet.java:883)
```

- With new design, we fail in `executeQuery` itself and throw `TableNotFoundException` when validating timestamps.
```
org.apache.phoenix.schema.TableNotFoundException: ERROR 1012 (42M03): Table undefined.
	at org.apache.phoenix.exception.SQLExceptionCode$9.newException(SQLExceptionCode.java:239) ~[classes/:?]
	at org.apache.phoenix.exception.SQLExceptionInfo.buildException(SQLExceptionInfo.java:229) ~[classes/:?]
	at org.apache.phoenix.util.ClientUtil.parseRemoteException(ClientUtil.java:115) ~[classes/:?]
	at org.apache.phoenix.util.ClientUtil.parseServerExceptionOrNull(ClientUtil.java:80) ~[classes/:?]
	at org.apache.phoenix.util.ClientUtil.parseServerException(ClientUtil.java:66) ~[classes/:?]
	at org.apache.phoenix.util.ValidateLastDDLTimestampUtil.validateLastDDLTimestamp(ValidateLastDDLTimestampUtil.java:112) ~[classes/:?]
	at org.apache.phoenix.jdbc.PhoenixStatement$1.call(PhoenixStatement.java:390) ~[classes/:?]
```

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java (hunk 5 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: We do not invalidate cache when we drop indexes on the server side as part of a drop column statement.

Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: Will make that design change as part of https://issues.apache.org/jira/browse/PHOENIX-7305 and revert this part of the test change.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java (hunk 10 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:1177]: Fixed the problem in https://issues.apache.org/jira/browse/PHOENIX-7307
Will revert this change

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java (hunk 5 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableWithViewsIT.java:1152]: Reverted

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java (hunk 10 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java]: reverted

File: phoenix-core-client/src/main/java/org/apache/phoenix/schema/MetaDataClient.java [hunk 18]
```diff
@@ -1776,7 +1778,18 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
             return buildIndexAtTimeStamp(table, statement.getTable());
         }
 
-        return buildIndex(table, tableRef);
+        MutationState state = buildIndex(table, tableRef);
+        // If client is validating LAST_DDL_TIMESTAMPS, parent's last_ddl_timestamp changed
+        // so remove it from client's cache. It will be refreshed when table is accessed next time.
+        boolean lastDDLTimestampValidationEnabled =
```
Comment (shahrs87) [inline phoenix-core-client/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:1784]: Do you want to use [ValidateLastDDLTimestampUtil#getValidateLastDdlTimestampEnabled](https://github.com/apache/phoenix/blob/PHOENIX-6883-feature/phoenix-core-client/src/main/java/org/apache/phoenix/util/ValidateLastDDLTimestampUtil.java#L73C27-L73C61) method instead?

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java [hunk 19]
```diff
@@ -1276,14 +1281,24 @@ public void testMetricsWithIndexUsage() throws Exception {
...
+                boolean lastDDLTimestampValidationEnabled =
+                        conn.unwrap(PhoenixConnection.class).getQueryServices().getProps().getBoolean(
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/PhoenixTableLevelMetricsIT.java:1287]: Do you want to use [ValidateLastDDLTimestampUtil#getValidateLastDdlTimestampEnabled](https://github.com/apache/phoenix/blob/PHOENIX-6883-feature/phoenix-core-client/src/main/java/org/apache/phoenix/util/ValidateLastDDLTimestampUtil.java#L73C27-L73C61) method instead?

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java [hunk 20]
```diff
@@ -152,12 +156,19 @@ public void testGlobalClientCacheMetrics() throws Exception {
         simulateWorkload("testGlobalClientCacheMetrics", numTables, numThreads, numMaxDML);
 
         // only 1 miss when the table is created
+        int numExpectedMisses = 1;
+        if (QueryServicesOptions.DEFAULT_LAST_DDL_TIMESTAMP_VALIDATION_ENABLED) {
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:160]: Why not to get the last ddl timestamp validation enabled flag from the connection properties?

File: phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheConnectionLevelPropIT.java [hunk 21]
```diff
@@ -61,6 +62,9 @@ public class UpdateCacheConnectionLevelPropIT extends ParallelStatsDisabledIT {
     private static Connection conn1;
     private static Connection conn2;
     private static ConnectionQueryServices spyForConn2;
+    private boolean isLastDDLTimestampValidationEnabled
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheConnectionLevelPropIT.java:65]: Why not to get the last ddl timestamp validation enabled flag from the connection properties?

File: phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheIT.java [hunk 22]
```diff
@@ -69,6 +70,10 @@
 @Category(ParallelStatsDisabledTest.class)
 public class UpdateCacheIT extends ParallelStatsDisabledIT {
 
+    private boolean isLastDDLTimestampValidationEnabled
+            = QueryServicesOptions.DEFAULT_LAST_DDL_TIMESTAMP_VALIDATION_ENABLED
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheIT.java:74]: Why not to get the last ddl timestamp validation enabled flag from the connection properties?

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java (hunk 20 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:160]: Not sure if there is a connection here in the test to use?

File: phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheConnectionLevelPropIT.java (hunk 21 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheConnectionLevelPropIT.java:65]: Not sure if there is a connection here in the test to use?

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java (hunk 20 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:160]: @shahrs87 I have made the change in MetaDataClient.java as that is production code. For the test code, maybe it is not required? Also in some tests, we do not have a connection object which can be used directly for `getValidateLastDdlTimestampEnabled`

Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:160]: I will make the change in tests to get the value from the config object in BaseTest.

File: phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java [hunk 23]
```diff
@@ -51,6 +51,10 @@ public class MetaDataCachingIT extends BaseTest {
     private static final Logger LOGGER = LoggerFactory.getLogger(MetaDataCachingIT.class);
     private final Random RAND = new Random(11);
 
+    private boolean isLastDDLTimestampValidationEnabled = config.getBoolean(
```
Comment (shahrs87) [inline phoenix-core/src/it/java/org/apache/phoenix/query/MetaDataCachingIT.java:54]: Instead of duplicating this line in 3 different tests, do you mind creating a util method in `ValidateLastDDLTimestampUtil` something like `getValidateLastDdlTimestampEnabled(Configuration config)` similar to `getValidateLastDdlTimestampEnabled(PhoenixConnection connection)`

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/FromCompiler.java [hunk 24]
```diff
@@ -297,7 +297,9 @@ public static ColumnResolver getIndexResolver(SingleTableStatement statement,
 
     public static ColumnResolver getResolver(SingleTableStatement statement, PhoenixConnection connection, Map<String, UDFParseNode> udfParseNodes)
             throws SQLException {
-        SingleTableColumnResolver visitor = new SingleTableColumnResolver(connection, statement.getTable(), true, 0, udfParseNodes);
+        SingleTableColumnResolver visitor
```
Comment (shahrs87) [inline phoenix-core-client/src/main/java/org/apache/phoenix/compile/FromCompiler.java:300]: I see that we are calling this resolver from `MetaDataClient#createIndex `and `CreateIndexCompiler#compile` method. So we want to bypass cache for createIndex and resolve table always. So lets rename this method to `getResolverForCreateIndex` and add a comment on why we are setting `alwaysHitServer` to true.

File: phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheIT.java (hunk 22 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/rpc/UpdateCacheIT.java]: will use config

**Review by shahrs87 — APPROVED** [review]
Other than 2 minor comments, +1 ltgm. Thank you @palashc  for the change.

**Comment by palashc** [comment]
@shahrs87 Latest test report - https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1859/22/testReport/
`UpsertSelectIT.testUpsertSelectForAgg` passes locally.

**Review by shahrs87 — APPROVED** [review]
